src/notebooks/controllers/ipywidgets/scriptSourceProvider/cdnWidgetScriptSourceProvider.ts

```diff
@@ -38,26 +42,34 @@ function moduleNameToCDNUrl(cdn: string, moduleName: string, moduleVersion: stri
         fileName = moduleName.substr(index + 1);
         packageName = moduleName.substr(0, index);
     }
-    if (cdn === jsdelivrUrl) {
-        // Js Delivr doesn't support ^ in the version. It needs an exact version
-        if (moduleVersion.startsWith('^')) {
+    if (cdnBaseUrl === jsdelivrUrl) {
+        cdn = {
+            url: cdnBaseUrl,
+            // Js Delivr doesn't support ^ in the version. It needs an exact version
+            requiresExactVersion: true,
+            // Js Delivr also needs the .js file on the end.
+            requiresExtension: true
+        };
+    }
+    if (cdn instanceof Object) {
+        if (cdn.requiresExactVersion && moduleVersion.startsWith('^')) {
```

Comment:
thanks for this PR, how about we change the format of the package json to just strings, `jsdelivr.com` and `unpkg.com` and for these we look at the value and hard code the requireExactVersion and requiresExtension.
If its something else, then we leave it as some default.

However this might not work for mirror jsdeliv sites. then again, we haven't had any requests for mirrored sites, hence I'm inclined to avoid having a more complex package.json and go with just strings.

What to you think?
Does this work for the site you're using?

Reply:
Thank you for taking a look!

> Does this work for the site you're using?

No, currently the CDN I target only supports `jsdelivr.com` format - but I am checking if this can be changed. Out of big CDNs `cdnjs.com` also requires an exact version and `.js` suffix (and compared to `jsdelivr.com` does not have redirects at all), although to support `cdnjs` the format should also include `/` instead of `@` for version specifier.

Another option would be giving some kind of a format string, like:

```
https://cdnjs.cloudflare.com/ajax/libs/<library>/<version>/<file>
```

This would keep the `package.json` to accept only strings, while allowing for a degree of customization. What do you think?